src/rocker/core.py

```diff
@@ -58,6 +63,24 @@ def validate_environment(self, cliargs):
         necessary resources are available, like hardware."""
         pass
 
+    @staticmethod
+    def preceding_extensions() -> typing.Set[str]:
+        """
+        Optional extensions. This merely ensures the preceding
+        extensions are applied before this extension when applying
+        snippets and arguments if they are present.
+        """
+        return set()
+
+    @staticmethod
+    def required_extensions() -> typing.Set[str]:
+        """
+        Ensures the specified extensions are present and combined with
+        this extension. In addition, it orders the application of
```

Comment:
Since we have an ordering constraint option it seems that we don't have to combine this entry to be both required and ordering.

Reply:
The way I currently have it implemented
- required extensions must be loaded (and they must be loaded before)
- preceding extensions are optional, but if they are loaded, they must be loaded before

Unless I misunderstood, I think we would need both options

Reply:
And maybe the confusion could be cleared up with better method_names. Maybe `required` and `optional`?

Reply:
>  It also allows you to specify extensions you want to be loaded after if they happen to be present, but not required.

I'm not sure how these two options provide this. 

I'm coming from the point of view that all other plugins are optional. But some need to come first. And the requirements is that they are there, but is independent of the order. Which is why I'm pushing back on this coupling "required extensions must be loaded (and they must be loaded before)"

Foo is required and needs to come first
If Bar exists it should come first

you can say 
preceds Foo, Bar
requires Foo 

Or as you've formulated it
preceds Bar
requires+follows Foo

I find this less intuitive. In the first case. You assert that the required ones are present(or inject). And then you sort based on the precedes for the ones present. The sorting and the requiring are independent. 

Keeping the sorting in a separate list makes reviewing the sorting expectations much easier. 

For naming I would actually suggest something more like `invoke_after` and `requires`

Reply:
Yep I'm now in agreement. Updated.